Update createmultisig RPC to support segwit

ajtowns opened the pull request:
Adds an "address_type" parameter that accepts "legacy", "p2sh-segwit", and "bech32" to choose the type of address created. Defaults to "legacy" rather than the value of the `-address-type` option for backwards compatibility.

As part of implementing this, OutputType is moved from wallet into its own module, and `AddAndGetDestinationForScript` is changed to apply to a `CKeyStore` rather than a wallet, and to invoke `keystore.AddCScript(script)` itself rather than expecting the caller to have done that.

Fixes #12502

ajtowns committed:
segwit support for createmultisig RPC

ajtowns committed:
[tests] functional test for createmultisig RPC

instagibbs commented:
I think this would resolve https://github.com/bitcoin/bitcoin/issues/12502

Thanks for working on this!

jonasschnelli commented:
Nice!
Concept ACK.

sipa commented:
@luke-jr The existing "legacy" address type (argument to `-addresstype` command line option) causes P2SH-multisig addresses for the multisig RPC. "p2sh-segwit" will cause P2SH-P2WSH-multisig addresses, and "bech32" will cause P2WSH-multisig.

ajtowns commented:
Ping @instagibbs @jonasschnelli ; Care to review code?

instagibbs commented on src/outputtype.cpp:
```diff
@@ -75,6 +75,8 @@ std::vector<CTxDestination> GetAllDestinationsForKey(const CPubKey& key)
 
 CTxDestination AddAndGetDestinationForScript(CKeyStore& keystore, const CScript& script, OutputType type)
 {
+    keystore.AddCScript(script);
```
is this a bugfix?

ajtowns replied:
It moves the `AddCScript()` call from the function invoking `AddAndGetDestinationForScript()` into `AAGDFS()`; see corresponding change in src/wallet/rpcwallet.cpp

instagibbs commented on src/rpc/misc.cpp:
```diff
@@ -134,14 +136,22 @@ static UniValue createmultisig(const JSONRPCRequest& request)
-    CScriptID innerID(inner);
+    const CScript inner = CreateMultisigRedeemscript(required, pubkeys);
+    CBasicKeyStore keystore;
+    const CTxDestination dest = AddAndGetDestinationForScript(keystore, inner, output_type);
```
I didn't think this rpc call changed wallet state previously?

ajtowns replied:
This shouldn't change wallet state either -- keystore is a dummy variable that's just declared here and discarded. The functional tests don't work without wallet support compiled in, but I think I gave it a quick test by hand and it was okay.

sipa commented on src/rpc/misc.cpp:
```diff
@@ -134,14 +136,22 @@ static UniValue createmultisig(const JSONRPCRequest& request)
     UniValue result(UniValue::VOBJ);
-    result.pushKV("address", EncodeDestination(innerID));
+    result.pushKV("address", EncodeDestination(dest));
     result.pushKV("redeemScript", HexStr(inner.begin(), inner.end()));
```
This field becomes essentially useless for new address types. I'm not sure what to do about that.

achow101 replied:
It could be changed to `script` and have `redeemScript` behind a deprecation switch.

ajtowns committed:
Add outputtype module

Moves OutputType into its own module

ajtowns committed:
Move AddAndGetDestinationForScript from wallet to outputype module

Makes AddAndGetDestinationForScript use a generic CKeyStore rather than
the wallet, and makes it always add the script to the keystore, rather
than only adding related (redeem) scripts.

achow101 commented:
utACK f40b3b82dfe873dd55ee24f4d6dec5d43756260a

laanwj commented:
utACK f40b3b82dfe873dd55ee24f4d6dec5d43756260a

sipa commented:
utACK f40b3b82dfe873dd55ee24f4d6dec5d43756260a.
